core/src/main/kotlin/Formats/MarkdownFormatService.kt

```diff
@@ -44,44 +118,57 @@ open class MarkdownFormatService(locationService: LocationService,
     }
 
     override fun formatLink(text: String, href: String): String {
-        return "[$text]($href)"
+        return """[${text.replace("[", "\\[").replace("]","\\]")}](${href})"""
     }
 
     override fun appendLine(to: StringBuilder, text: String) {
-        to.appendln(text)
+        to.appendWithAdjustedLF(text)
+    }
+
+    override fun appendList(to: StringBuilder, body: () -> Unit) {
+        to.append(allowedLf().requiredWhitespace)
+        changeLf(LfPriority.HtmlBr) { body() }
     }
 
     override fun appendAnchor(to: StringBuilder, anchor: String) {
         // no anchors in Markdown
     }
 
     override fun appendParagraph(to: StringBuilder, text: String) {
-        to.appendln()
-        to.appendln(text)
-        to.appendln()
+        to.append(allowedLf().paragraph.start)
+        to.append(text)
+        to.append(allowedLf().paragraph.end)
     }
 
     override fun appendHeader(to: StringBuilder, text: String, level: Int) {
-        appendLine(to)
-        appendLine(to, "${"#".repeat(level)} $text")
-        appendLine(to)
-    }
-
-    override fun appendBlockCode(to: StringBuilder, lines: List<String>, language: String) {
-        appendLine(to)
-        to.appendln(if (language.isEmpty()) "```" else "``` $language")
-        to.appendln(lines.joinToString("\n"))
-        to.appendln("```")
-        appendLine(to)
-    }
-
-    override fun appendTable(to: StringBuilder, body: () -> Unit) {
-        to.appendln()
-        body()
-        to.appendln()
-    }
-
-    override fun appendTableHeader(to: StringBuilder, body: () -> Unit) {
+        to.append(allowedLf().requiredWhitespace)
+        to.appendWithAdjustedLF("${"#".repeat(level)} $text")
+    }
+
+    override fun appendBlockCode(to: StringBuilder, language: String, bodyAsLines: ()->List<String>) {
+        to.append(allowedLf().requiredWhitespace)
+
+        val codeTags = allowedLf().inlineCode
+        if (codeTags.start.startsWith('`')) {
+            changeLf(LfPriority.Code) {
+                // this does not adjust line feeds, a code block in a list or table will just break formatting
+                to.appendln(if (language.isEmpty()) "```" else "``` $language")
+                // use 4 zero width spaces which act as a code fence escaping any use of ``` within the code block
+                to.appendln(bodyAsLines().map { "​\u200B\u200B\u200B\u200B$it" }.joinToString("\n"))
```

Comment:
Few markdown renderers support headerless tables, and common mark the parser you are using doesn't either.  It is currently being discussed how headerless tables would be in markdown and there is no consensus.  The table format that was there before was the rare unsupported case of having no headers and no `---|---|---` row.